include/bout/region.hxx

```diff
@@ -0,0 +1,236 @@
+/**************************************************************************
+ * Copyright 2010 B.D.Dudson, S.Farley, M.V.Umansky, X.Q.Xu
+ *
+ * Region class by D. Dickinson 2018
+ *
+ * Contact: Ben Dudson, bd512@york.ac.uk
+ * 
+ * This file is part of BOUT++.
+ *
+ * BOUT++ is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * BOUT++ is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with BOUT++.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **************************************************************************/
+
+#ifndef __REGION_H__
+#define __REGION_H__
+
+#include <type_traits>
+#include <vector>
+#include <utility>
+
+#include "bout/openmpwrap.hxx"
+
+/*
+ * The MAXREGIONBLOCKSIZE value can be tuned to try to optimise
+ * performance on specific hardware. It determines what the largest
+ * contiguous block size can be. As we hope the compiler will vectorise
+ * the access to these contiguous blocks, the optimal MAXREGIONBLOCKSIZE 
+ * is likely related to the vector size etc.
+ */
+#ifndef MAXREGIONBLOCKSIZE
+#define MAXREGIONBLOCKSIZE 64
+#endif
+#if MAXREGIONBLOCKSIZE <= 0
+#error "In region.hxx must set MAXREGIONBLOCKSIZE > 0"
+#endif
+
+//The following loops vectorise well well index is type int, needs testing
+//with current approach where index is ind2D or ind3D
+#define BLOCK_REGION_LOOP_SERIAL(region, index, ...)			\
+  {const auto blocks = region.getBlocks();				\
+  for(auto block = blocks.begin(); block < blocks.end() ; ++block){ \
+  for(auto index = (*block).first ; index <= (*block).second; ++index){ \
+  __VA_ARGS__ \
+    }}}
+
+#define BLOCK_REGION_LOOP(region, index, ...)			\
+  {const auto blocks = region.getBlocks();			\
+  BOUT_OMP(parallel for)				    \
+  for(auto block = blocks.begin(); block < blocks.end() ; ++block){ \
+  for(auto index = (*block).first ; index <= (*block).second; ++index){ \
+  __VA_ARGS__ \
+    }}}
+
+class specificInd{
+public:
+  int ind;
+  specificInd(){};
+  specificInd(int i): ind(i){};
+  specificInd& operator++(){++ind;return *this;}
+  specificInd& operator++(int){++ind;return *this;}
```

Comment:
This is extremely dangerous!
Rather than doing it wrong - throw an exception?

Reply:
Could you expand, not sure which line(s) you're referring to?

Reply:
line 71.
The postfix operator is not a prefix operator.
If I write `i++` I certainly don't expect to get `++i`.
This is an example that would be currently broken:
```
for (;i != i.end;){
  auto fu=in[i];
  r[i++]=fu;
}
```
`i++` should return the non-incremented value - not the incremented one.
See https://stackoverflow.com/questions/3846296/how-to-overload-the-operator-in-two-different-ways-for-postfix-a-and-prefix
https://stackoverflow.com/questions/24853/what-is-the-difference-between-i-and-i

Reply:
Ah yes, thanks! I should note that this error is already in the code inside `include/bout/sys/range.hxx` (where I quickly grepped for the syntax for this as I didn't have internet at the time)